### PR TITLE
HPCC-21029 Improve AsyncFor code

### PR DESCRIPTION
**Pull request by richardkchapman**
Executes the last cycle on the calling thread, in the common simple case.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Added unit tests, regression suite

### CODE REVIEW

File: system/jlib/jthread.cpp [hunk 1]
```diff
@@ -702,22 +702,47 @@ void CAsyncFor::For(unsigned num,unsigned maxatonce,bool abortFollowingException
...
+                started.item(idx).join();
+            }
         }
-        ForEachItemIn(idx, started)
+        else
         {
-            started.item(idx).join();
+            // Common case of execute all at once can be optimized a little
+            // Note that shuffle and abortFollowingException are meaningless when executing all at once
+            dbgassertex(!shuffled);
```
Comment (ghalliday) [inline system/jlib/jthread.cpp]: not sure this is a good idea.  Some general purpose code e.g.

    afor.For(eps.ordinality(), 10, false, true);

will trigger the assert if the list is too short

Comment (richardkchapman) [reply system/jlib/jthread.cpp]: True. Unlikely, but I will remove the asserts

File: system/jlib/jthread.cpp [hunk 2]
```diff
@@ -702,22 +702,45 @@ void CAsyncFor::For(unsigned num,unsigned maxatonce,bool abortFollowingException
...
+            IArrayOf<Thread> started;
+            started.ensure(num);
+            for (i=0;i<num-1;i++)
+            {
+                Owned<Thread> thread = new cdothread(this,i,ready,&errmutex,e);
+                thread->start();
+                started.append(*thread.getClear());
+            }
+            Do(num-1);
+            ready.wait(); // Waiting on a single semaphore should help reduce context switches, even though we still need to join below
```
Comment (ghalliday) [inline system/jlib/jthread.cpp]: This isn't doing what you hope for - the reinit will have set it to a number > 0, so the wait will return immediately.

You cannot supply a -ve value to sem_init, so I'm not sure you can do what you are aiming for.

**Comment by richardkchapman** [comment]
@ghalliday

**Pull request merged**